Throwing an exception when setting hidden_envoy_deprecated fields

adisuissa opened the pull request:
Throwing an exception when setting hidden_envoy_deprecated fields and adding tests
Disabling direct setting of hidden_envoy_deprecated fields, while still allowing the setting by an upgraded protobuf message.
Fixes: #9496

Risk Level: Medium - in cases where operators were using deprecated fields and they weren't supposed to, an exception will be thrown.
Testing: Added a general test for deprecated.proto, and a specific test for EDS.

Signed-off-by: Adi Suissa-Peleg <adip@google.com>

adisuissa commented:
Added tests for all the cases, and converted the integration tests to use binary protobuf bootstrap config, and runtime override for xDS updates.
@htuch I think all comments were addressed, but let me know if there's anything else that you think can improve this.

htuch commented on test/common/router/config_impl_test.cc:
```diff
@@ -136,6 +136,14 @@ Http::TestRequestHeaderMapImpl genHeaders(const std::string& host, const std::st
+  envoy::config::route::v3::RouteConfiguration route_config;
+  TestUtility::loadFromYaml(yaml, route_config, true);
+  TestUtility::validate(route_config);
+  return route_config;
```
Is this intended to be just clearer or functionally different? I think both versions will boost, since `do_boosting` is true by default.

adisuissa replied:
Previously I've implemented v3 loading differently, but I agree it isn't needed.
I've updated the function name (`parseRouteConfigurationFromYaml`, removed V2/3), and made sure it loads V2 and V3 correctly.

htuch commented on test/common/router/config_impl_test.cc:
```diff
@@ -6265,7 +6273,7 @@ name: AllRedirects
   )EOF";
 
-  TestConfigImpl config(parseRouteConfigurationFromV2Yaml(RouteDynPathRedirect), factory_context_,
+  TestConfigImpl config(parseRouteConfigurationFromV3Yaml(RouteDynPathRedirect), factory_context_,
```
Are these all v3? I'm guessing some haven't been updated. Doing the parse for v3 with `do_boosting = false` would probably show.

adisuissa replied:
You are correct, I've updated the loading function name

adisuissa commented on source/common/config/version_converter.cc:
```diff
@@ -218,5 +216,7 @@ void VersionUtil::scrubHiddenEnvoyDeprecated(Protobuf::Message& message) {
   ProtobufMessage::traverseMutableMessage(proto_visitor, message, nullptr);
 }
 
+const char VersionUtil::DeprecatedFieldShadowPrefix[] = "hidden_envoy_deprecated_";
```
Added usage here:
https://github.com/envoyproxy/envoy/blob/fbc6c97ccebec791ff459ed0582400a159cb6a16/source/common/protobuf/utility.cc#L467

adisuissa commented on source/common/protobuf/utility.cc:
```diff
@@ -347,13 +347,38 @@ void MessageUtil::loadFromFile(const std::string& path, Protobuf::Message& messa
+        }
+      }
+      if (message_version == MessageVersion::LATEST_VERSION) {
+        throw EnvoyException("Unable to parse file \"" + path + "\" as a binary protobuf (type " +
```
Cannot be reached... Updated code.

adisuissa commented on source/common/protobuf/utility.cc:
```diff
@@ -439,6 +464,19 @@ class UnexpectedFieldProtoVisitor : public ProtobufMessage::ConstProtoVisitor {
+        // unless it is part of an explicit test that needs access to the deprecated field
+        if (!runtime_ ||
+            !runtime_->snapshot().deprecatedFeatureEnabled(
+                absl::StrCat("envoy.deprecated_features:", field.full_name()), false)) {
```
Let me know if this can be clearer (I thought the comment made sense,  but it might need to be improved)

htuch replied:
It's probably worth leaving a note such as "When we enable universal runtime deprecation override, this must not be applied here, since we only want point field overrides for test.".

adisuissa commented on source/common/protobuf/utility.h:
```diff
@@ -362,6 +362,13 @@ class MessageUtil {
+   * @param message message of type type.googleapis.com/google.protobuf.Message.
+   * @return std::string of the Protobuf binary object.
+   */
+  static std::string getProtobufBinaryStringFromMessage(const Protobuf::Message& message);
```
Moved.

adisuissa commented on test/common/protobuf/utility_test.cc:
```diff
@@ -1674,6 +1681,49 @@ TEST_P(DeprecatedFieldsTest, DEPRECATED_FEATURE_TEST(FatalEnum)) {
       checkForDeprecation(base));
 }
 
+TEST_P(DeprecatedFieldsTest, ManualDeprecatdFieldAddition) {
```
Added

adisuissa commented on test/common/protobuf/utility_test.cc:
```diff
@@ -1674,6 +1681,49 @@ TEST_P(DeprecatedFieldsTest, DEPRECATED_FEATURE_TEST(FatalEnum)) {
+TEST_P(DeprecatedFieldsTest, ManualDeprecatdFieldAddition) {
+  // Create a base message and insert a deprecated field. When upgrading the
+  // deprecated field should be set as deprecated, and a warning should be logged
+  envoy::test::deprecation_test::Base baseShouldWarn =
```
Done

adisuissa commented on test/common/protobuf/utility_test.cc:
```diff
@@ -1674,6 +1681,49 @@ TEST_P(DeprecatedFieldsTest, DEPRECATED_FEATURE_TEST(FatalEnum)) {
+
+  // Create an upgraded message and insert a deprecated field. This is a bypass
+  // of the upgrading procedure validation, and should fail
+  envoy::test::deprecation_test::UpgradedBase baseShouldFail =
```
Done

adisuissa commented on source/common/protobuf/utility.cc:
```diff
@@ -439,6 +464,19 @@ class UnexpectedFieldProtoVisitor : public ProtobufMessage::ConstProtoVisitor {
+            !runtime_->snapshot().deprecatedFeatureEnabled(
+                absl::StrCat("envoy.deprecated_features:", field.full_name()), false)) {
+          const std::string fatal_error = absl::StrCat(
+              "Illegal use of deprecated option '", field.full_name(), "' from file ", filename,
```
Done.

adisuissa commented:
@htuch I've addressed the comments. I'm seeing a recurring problem with CI. I get the following error message:
```
Format check failed, try apply following patch to fix:
diff --git a/test/config_test/BUILD b/test/config_test/BUILD
index 154dc3ec4..b5361aa49 100644
--- a/test/config_test/BUILD
+++ b/test/config_test/BUILD
@@ -62,7 +62,5 @@ envoy_cc_test(
         "//test/test_common:environment_lib",
         "//test/test_common:logging_lib",
         "//test/test_common:utility_lib",
-        "@envoy_api//envoy/config/bootstrap/v2:pkg_cc_proto",
-        "@envoy_api//envoy/config/bootstrap/v3:pkg_cc_proto",
     ],
 )
```

However, when I remove the lines from the code, and try to push it, I get:
```
  Checking format for test/config_test/BUILD - ERROR: From ./test/config_test/BUILD
ERROR: envoy_build_fixer check failed for file: ./test/config_test/BUILD
ERROR:   ./test/config_test/BUILD:64
ERROR: check format failed. run 'tools/code_format/check_format.py fix'
```
Which suggests to add the two lines back.
Any ideas?

htuch commented:
@adisuissa are you missing `deprecated_configs_test.cc` from the patch you push to GH?

htuch commented on test/common/protobuf/utility_test.cc:
```diff
@@ -1676,6 +1683,51 @@ TEST_P(DeprecatedFieldsTest, DEPRECATED_FEATURE_TEST(FatalEnum)) {
 
+// Verify that direct use of a hidden_envoy_deprecated field fails, but upgrade
+// succeeds
+TEST_P(DeprecatedFieldsTest, ManualDeprecatdFieldAddition) {
```
Nit: s/Deprecatd/Deprecated/g

adisuissa replied:
Done

htuch commented on test/common/upstream/eds_test.cc:
```diff
@@ -1873,6 +1873,34 @@ TEST_F(EdsAssignmentTimeoutTest, AssignmentLeaseExpired) {
+  Protobuf::RepeatedPtrField<ProtobufWkt::Any> resources;
+  resources.Add()->PackFrom(cluster_load_assignment);
+  EXPECT_THROW_WITH_REGEX(eds_callbacks_->onConfigUpdate(resources, ""), ProtoValidationException,
+                          "Illegal use of deprecated V2 option "
```
Nit: I think clarifying the message to say "Illegal use of internal hidden_envoy_deprecated_ field" might help.

adisuissa replied:
Done

adisuissa commented:
> @adisuissa are you missing `deprecated_configs_test.cc` from the patch you push to GH?

derp... adding the file

adisuissa commented:
@htuch Thanks for noticing the missing file. I've addressed the comments.

htuch commented on test/config_test/deprecated_configs_test.cc:
```diff
@@ -0,0 +1,239 @@
+                "'envoy.config.core.v3.Node.hidden_envoy_deprecated_build_version'"));
+}
+
+// A deprecated field can be used in previous version yaml and upgraded.
```
Nit: JSON.

adisuissa replied:
Done

adisuissa commented:
Seems that one of the problems is in the `envoy-filter-example` repo.
I've submitted [PR 127](https://github.com/envoyproxy/envoy-filter-example/pull/127) in that repo, that should address the issue.

htuch commented:
@adisuissa I've merged your envoy-filter-example PR.
/wait

adisuissa commented:
/retest

adisuissa commented:
/retest

htuch commented:
ASAN failure is an unrelated timeout, I'll merge this.
